lib/TableGen/Operations.cpp

```diff
@@ -570,24 +626,48 @@ void BuilderMethod::emitDefinition(raw_ostream &out, FmtContext &fmt,
     }
 
     fmt.addSubst("args", symbols.chooseName({"args", "arguments"}));
-    out << tgfmt("::llvm::Value* const $args[] = {\n", &fmt);
+    out << tgfmt("std::vector<::llvm::Value*> $args = {\n", &fmt);
+    const NamedValue *varArg = nullptr;
+
+    size_t ArgIdx = 0;
     for (const auto& [opArg, builderArg, attrType]
               : llvm::zip_first(fullArguments, opArgBuilderArgs, m_attrTypes)) {
+      if (ArgIdx > 0 && !opArg.type->isVarArgList())
+        out << ",\n";
+
       if (auto* attr = dyn_cast<Attr>(opArg.type)) {
         out << tgfmt(attr->getToLlvmValue(), &fmt, builderArg.name, attrType);
       } else if (opArg.type->isTypeArg()) {
         out << tgfmt("::llvm::PoisonValue::get($0)", &fmt, builderArg.name);
+      } else if (opArg.type->isVarArgList()) {
+        varArg = &opArg;
+        // Assume no other arguments follow the variadic argument.
+        --ArgIdx;
+        break;
       } else {
         out << builderArg.name;
       }
-      out << ",\n";
+
+      ++ArgIdx;
     }
 
+    std::string varArgInitializer = "";
+    if (varArg) {
+      varArgInitializer = tgfmt(R"(
+        ::llvm::for_each($0, [&](::llvm::Value *varArg) {
+          $args.push_back(varArg);
+        });
```

Comment:
For a SmallVector, this can be `$args.append($0.begin(), $0.end())`, which reserves the needed size upfront.